chore: add offers and permissions during migration import

kian99 opened the pull request:
## Description
This PR makes a few tweaks to the `Import` step of model migration, namely:
- Creates offers in JIMM's database alongside model creation.
- Adds permissions for the new model owner to the model and adds the correct permissions for any application offers.
- Ensures permissions are removed during `Abort`.

I also refactored the logic in `Import` to encapsulate it, and avoid leaking methods that are specific to the `Import` method to the rest of the Juju manager.

Fixes [JUJU-8192](https://warthogs.atlassian.net/browse/JUJU-8192)

## Engineering checklist
- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

[JUJU-8192]: https://warthogs.atlassian.net/browse/JUJU-8192?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kian99 committed:
chore: add offers and permissions during migration import

SimoneDutto commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -69,13 +70,17 @@ func (j *JujuManager) AbortMigration(ctx context.Context, user *openfga.User, mo
 	if err != nil {
 		zapctx.Error(ctx, "failed to delete incoming model migration", zap.Error(err), zap.String("modelUUID", modelUUID))
 	}
+	err = j.OpenFGAClient.RemoveModel(ctx, model.ResourceTag())
```
i think we need this as well in the Cleanup

kian99 replied:
I don't think so, if there are dangling tuples i.e. tuples that reference a resource that doesn't exist, we have an existing cleanup routine to remove them.

SimoneDutto replied:
fair enough

alesstimec replied:
even if we don't remove the model from OpenFGA in this step all tuples referencing it should be removed by the cleanup routine

SimoneDutto commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -447,57 +491,115 @@ func (j *JujuManager) Import(ctx context.Context, user *openfga.User, serialized
+		for _, app := range description.Applications() {
+			for _, offer := range app.Offers() {
+				// construct the offer URL with the same logic as Juju (modelOwner, modelName, offerName, <blank-controller-name>)
+				offerURL := jujucrossmodel.MakeURL(description.Owner().Id(), modelNameStr, offer.OfferName(), "")
```
question: what's the forth argument?

kian99 replied:
The controller name

SimoneDutto replied:
I guess the controller_name must be specified only when you are consuming a cross-controller relations.
Which shouldn't be an issue because here we are just importing offers in JIMM's state

SimoneDutto commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -447,57 +491,115 @@ func (j *JujuManager) Import(ctx context.Context, user *openfga.User, serialized
+		appAndOffers.model = model
+
+		for _, app := range description.Applications() {
+			for _, offer := range app.Offers() {
```
just to make sure i understand:
we create the offer in JIMM and the target juju controller with the external identity, we will then:
- convert the local user to external in the discharge endpoint (already done)
- find a solution to map these offers for terraform as we've discussed yesterday

kian99 replied:
>we create the offer in JIMM and the target juju controller with the external identity, we will then:

Yes we will create the offer in JIMM. Juju will create the offer in its handling of the `Import` call (using the model owner as part of the offer URL).

On the other two points,
1. Yes, we will look up the user-mappings for the model in the discharger to convert local users to external users.
2. Yip, we'll have to handle that in https://warthogs.atlassian.net/browse/JUJU-8247

SimoneDutto commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -400,11 +405,22 @@ func (j *JujuManager) Activate(ctx context.Context, modelTag names.ModelTag, mig
+// This helps to keep the import logic and associated methods separate from the JujuManager
+// methods.
+type modelMigrationImport struct {
+	*JujuManager
```
i would prefer here to have an more specific interface here to better understand the methods of the juju manager used by the modelMigrationImport.
What do you think?

kian99 replied:
I can see that being useful. But it feels like a bit too much abstraction without the need for it right now (beyond some readability). We don't test this piece in isolation, we just test the exported Import method so I'm in favor of not adding an interface for it (at least for now).

SimoneDutto replied:
i don't like this encapsulation too much because:
- it's a little bit of an inception to call a juju manager method from inside a modelMigrationImport's method called itself from juju manager
- i don't like that we use model description from this struct that is being modified as a side effect of calling methods on this struct

But if the other reviewers are fine, i won't be opposed to it

ale8k commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -400,11 +405,22 @@ func (j *JujuManager) Activate(ctx context.Context, modelTag names.ModelTag, mig
+// modelMigrationImport is a struct that holds the JujuManager and the model description
+// that is being imported. It is used to encapsulate the logic for importing a model
+// from a serialized description.
+// This helps to keep the import logic and associated methods separate from the JujuManager
```
If we have to doc that we're making this for encapsulation, then we're likely doing something wrong I think. We shouldn't have to explicitly state we're creating a struct to encapsulate logic.

ale8k commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -413,32 +429,60 @@ func (j *JujuManager) Import(ctx context.Context, user *openfga.User, serialized
 		return errors.E(op, fmt.Errorf("failed to deserialize model description: %w", err))
 	}
+
+	importer := modelMigrationImport{
```
We hav the struct defined above, and then a JujuManager method, can we keep things in-order please? It makes it easier to follow if we have 

..juju manager method
..juju manager method
..modelMigrationImport struct
..modelMigrationImport method
..modelMigrationImport method

ale8k replied:
Also maybe rename it to modelMigrationImporter? We call it "importer" here

ale8k commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -400,11 +405,22 @@ func (j *JujuManager) Activate(ctx context.Context, modelTag names.ModelTag, mig
+// from a serialized description.
+// This helps to keep the import logic and associated methods separate from the JujuManager
+// methods.
+type modelMigrationImport struct {
```
Needs a constructor to protect us from a nil JujuManager

ale8k commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -413,32 +429,60 @@ func (j *JujuManager) Import(ctx context.Context, user *openfga.User, serialized
+// migrationImport imports resources from the model description into JIMM's state
+// and adds the necessary permissions for the model and application offers
+// before calling the import method on the target Juju controller.
+// Since `import` is a reserved word in Go, we use `migrationImport`.
```
?? You can't call it importMigration?

kian99 replied:
You can't call it `import` since it's a reseved keyword.

ale8k commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -447,57 +491,115 @@ func (j *JujuManager) Import(ctx context.Context, user *openfga.User, serialized
 	}
 
-	err := j.Database.GetCloudCredential(ctx, cloudCredential)
+	err := mmi.Database.GetCloudCredential(ctx, cloudCredential)
```
We're transitively using the database from the JujuManager with no guarantee the JujuManager nor Database is safe to use. I think I'd much prefer to see two interfaces, one for the JujuManager and one for the Database methods that will be used within the modelMigrationImporter. 

I think this could just be two functions and it'd read clearer, because really at the end of the day the signatures within the importer are:

```go
func (mmi *modelMigrationImport) addResourcePermissions(ctx context.Context, user *openfga.User, resources *importedResources, targetControllerTag names.ControllerTag) error
func (mmi *modelMigrationImport) importFromDescription(ctx context.Context, targetControllerID uint, description description.Model) (*importedResources, error)
```

So I'd be in favour of just two simple functions in this file and even remove the type "importedResources", because without validating (as in to check nil on offers & model is set correctly) it it's no different from returning the two values.

ale8k commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -413,32 +429,60 @@ func (j *JujuManager) Import(ctx context.Context, user *openfga.User, serialized
 	}
-	err = j.modifyModelDescription(modelDescription, incomingMigration.UserMapping)
+
+	err = mmi.modifyModelDescription(mmi.modelDescription, incomingMigration.UserMapping)
```
Calling through to JujuManager is confusing here I think (for a moment it looked like this would be a responsibility of the importer)

ale8k commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -447,57 +491,115 @@ func (j *JujuManager) Import(ctx context.Context, user *openfga.User, serialized
+	return &appAndOffers, nil
+}
+
+func (mmi *modelMigrationImport) addResourcePermissions(ctx context.Context, user *openfga.User, resources *importedResources, targetControllerTag names.ControllerTag) error {
```
For consistency maybe godoc this too? The name doesn't really read so clearly, that is, it doesn't say to me "set owner of model and relate app offers to this model / rename this func

ale8k reviewed (commented):
Needs to rereview and have a think... I'm not a fan of having a struct (the importer) with two methods that has the JujuManager inline and then uses those inline calls that are specific to model imports inside its own functions. I think just two funcs would read a lot clearer to be honest. 

TLDR; the separation here is more confusing than helpful.

Code wise everything seems fine to me though.

kian99 commented:
@SimoneDutto @ale8k Let me remove that struct. Will ping you guys when that's done.

kian99 committed:
chore: remove modelMigrationImport struct

kian99 commented:
@SimoneDutto @ale8k I've removed the modelMigrationImport struct so it should read more clearly now.

SimoneDutto reviewed (approved):
lgtm

alesstimec commented on internal/jimm/juju/migrationtarget.go:
```diff
@@ -413,25 +419,35 @@ func (j *JujuManager) Import(ctx context.Context, user *openfga.User, serialized
 
+	err = j.addModelAndOfferPermissions(ctx, user, model, offers)
+	if err != nil {
+		return errors.E(op, fmt.Errorf("failed to add resource permissions: %w", err))
```
now, this is a bit of a tricky situation.. we've imported the model into JIMM's state, but failed to set permissions. how do we get out of this situation? should we give users the tooling to re-set model and offer permissions? but the problem is that we don't have a record of who should have permission to them..

kian99 replied:
I think this might be okay, if this fails, the migration will be aborted and in JIMM's Abort handler we remove the model.

kian99 committed:
chore: explicitly use controller tag

When creating a model, the model's Controller ID needs to be present but our CreateModel function doesn't fetch the controller's details and place this on the model struct, as opposed to GetModel which does.
This means after model creation, if we want the controller tag, we need to either use the controller details that we are certain we have or run GetModel.
Here I've opted to just pass the controller details that we already have to `addModelAndOfferPermissions`.

alesstimec reviewed (approved):
LGTM.. but we need to think about the situation where we fail to set permissions to imported model/offers..
